# Udacity Project 3
# Wrangle and Analyze Data

## Table of Contents
<ul>
<li><a href="#intro">Introduction</a></li>
<li><a href="#data_gathering">Data Gathering Data</a></li>
<li><a href="#data_assessing">Data Assessing</a></li>
<li><a href="#data_cleaning">Data Cleaning</a></li>
</ul>

<a id='intro'></a>
# Introduction

Data wrangling is a core skill that everyone who works with data should be familiar with since so much of the world's data isn't clean. This project aims to to get insightful information about dog ratings from the twitter page WeRateDogs™, while demonstrating advanced data wrangling and visualization techniques using various Python libraries. 

<a id='data_gathering'></a>
# Data Gathering

In [1]:
#import all needed libraries
import pandas as pd
import requests
import tweepy as tp
import json
import numpy as np

### Reading a csv-file unsing Pandas and storing the data into DataFrame

In [2]:
#reading twitter-archive-enhanced.csv into df1
df1 = pd.read_csv('twitter-archive-enhanced.csv')
df1.head()

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  892420643555336193                    NaN                  NaN   
1  892177421306343426                    NaN                  NaN   
2  891815181378084864                    NaN                  NaN   
3  891689557279858688                    NaN                  NaN   
4  891327558926688256                    NaN                  NaN   

                   timestamp  \
0  2017-08-01 16:23:56 +0000   
1  2017-08-01 00:17:27 +0000   
2  2017-07-31 00:18:03 +0000   
3  2017-07-30 15:58:51 +0000   
4  2017-07-29 16:00:24 +0000   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   
1  <a href="http://twitter.com/download/iphone" r...   
2  <a href="http://twitter.com/download/iphone" r...   
3  <a href="http://twitter.com/download/iphone" r...   
4  <a href="http://twitter.com/download/iphone" r...   

                                                text  retweeted_status_id  \
0  This is Phineas. He's a mystical boy. Only eve...                  NaN   
1  This is Tilly. She's just checking pup on you....                  NaN   
2  This is Archie. He is a rare Norwegian Pouncin...                  NaN   
3  This is Darla. She commenced a snooze mid meal...                  NaN   
4  This is Franklin. He would like you to stop ca...                  NaN   

   retweeted_status_user_id retweeted_status_timestamp  \
0                       NaN                        NaN   
1                       NaN                        NaN   
2                       NaN                        NaN   
3                       NaN                        NaN   
4                       NaN                        NaN   

                                       expanded_urls  rating_numerator  \
0  https://twitter.com/dog_rates/status/892420643...                13   
1  https://twitter.com/dog_rates/status/892177421...                13   
2  https://twitter.com/dog_rates/status/891815181...                12   
3  https://twitter.com/dog_rates/status/891689557...                13   
4  https://twitter.com/dog_rates/status/891327558...                12   

   rating_denominator      name doggo floofer pupper puppo  
0                  10   Phineas  None    None   None  None  
1                  10     Tilly  None    None   None  None  
2                  10    Archie  None    None   None  None  
3                  10     Darla  None    None   None  None  
4                  10  Franklin  None    None   None  None

In [3]:
df1.shape

(2356, 17)

In [299]:
df_archive = df1.copy()

### Downloading the tsv-file from intenten using request library and storing the data into DataFrame

In [5]:
#download data from image-predictions.tsv
req_data = requests.get('https://d17h27t6h515a5.cloudfront.net/topher/2017/August/599fd2ad_image-predictions/image-predictions.tsv')

In [6]:
#writing data to image-predictions.tsv
with open('image-predictions.tsv', mode='wb') as file:
    file.write(req_data.content)

In [7]:
#storing data from image-predictions.tsv to df2
df2 = pd.read_csv('image-predictions.tsv', sep='\t')
df2.head()

tweet_id                                          jpg_url  \
0  666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1  666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2  666033412701032449  https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3  666044226329800704  https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4  666049248165822465  https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   

   img_num                      p1   p1_conf  p1_dog                  p2  \
0        1  Welsh_springer_spaniel  0.465074    True              collie   
1        1                 redbone  0.506826    True  miniature_pinscher   
2        1         German_shepherd  0.596461    True            malinois   
3        1     Rhodesian_ridgeback  0.408143    True             redbone   
4        1      miniature_pinscher  0.560311    True          Rottweiler   

    p2_conf  p2_dog                   p3   p3_conf  p3_dog  
0  0.156665    True    Shetland_sheepdog  0.061428    True  
1  0.074192    True  Rhodesian_ridgeback  0.072010    True  
2  0.138584    True           bloodhound  0.116197    True  
3  0.360687    True   miniature_pinscher  0.222752    True  
4  0.243682    True             Doberman  0.154629    True

In [8]:
df2.shape

(2075, 12)

In [374]:
df_image = df2.copy()

### Query the Twitter API for tweets in the Twitter archive and save JSON in a text file

In [10]:
#connecting to the twitter api
consumer_key = 'S42Hgv6BqUOQYzY47ZcCdO2gZ'
consumer_secret = 'RumHDN7i4LFKB1JKPmdkuMZnCzvLcfLjCyym0TcF0Ppsjz3lM7'
access_token = '989486218068090880-JGPGRJCSVpKZ5PLtKoRYL7VzMeB8wKL'
access_secret = '9rVfO5Keo46NNQAojGI9BJ5dtWsjznue1wVTeC8cuZVH1'

auth = tp.OAuthHandler(consumer_key, consumer_secret)
auth.set_access_token(access_token, access_secret)

api = tp.API(auth, wait_on_rate_limit = True, wait_on_rate_limit_notify = True)

In [11]:
#obtaining a list of dictionaries from the twitter api
#with informations about each tweets id, favorite count, and retweet count

df_list = []
errors = []
for tweet_id in df_archive['tweet_id']:
    try:
        tweet = api.get_status(tweet_id, tweet_mode='extended')
        df_list.append({'tweet_id': str(tweet.id),
                        'favorite_count': int(tweet.favorite_count),
                        'retweet_count': int(tweet.retweet_count)})
    except Exception as e:
        print(str(tweet_id) + " : " + str(e))
        errors.append(tweet_id)

888202515573088257 : [{'code': 144, 'message': 'No status found with that ID.'}]
873697596434513921 : [{'code': 144, 'message': 'No status found with that ID.'}]
872668790621863937 : [{'code': 144, 'message': 'No status found with that ID.'}]
872261713294495745 : [{'code': 144, 'message': 'No status found with that ID.'}]
869988702071779329 : [{'code': 144, 'message': 'No status found with that ID.'}]
866816280283807744 : [{'code': 144, 'message': 'No status found with that ID.'}]
861769973181624320 : [{'code': 144, 'message': 'No status found with that ID.'}]
856602993587888130 : [{'code': 144, 'message': 'No status found with that ID.'}]
851953902622658560 : [{'code': 144, 'message': 'No status found with that ID.'}]
845459076796616705 : [{'code': 144, 'message': 'No status found with that ID.'}]
844704788403113984 : [{'code': 144, 'message': 'No status found with that ID.'}]
842892208864923648 : [{'code': 144, 'message': 'No status found with that ID.'}]
837366284874571778 : [{'code

In [12]:
#saving obtained data into a file
with open('tweet_json.txt', 'w') as outfile:  
    json.dump(df_list, outfile)

In [13]:
#reading data from file into df3
with open('tweet_json.txt', 'r') as file:
    df3 = pd.DataFrame(json.load(file), columns=['tweet_id', 'favorite_count', 'retweet_count'])

In [14]:
df3.head()

tweet_id  favorite_count  retweet_count
0  892420643555336193           36297           7725
1  892177421306343426           31301           5718
2  891815181378084864           23570           3786
3  891689557279858688           39598           7888
4  891327558926688256           37789           8506

In [15]:
df3.shape

(2331, 3)

In [16]:
df_tweet = df3.copy()

### Gathering Data Summary

Gathering the data was the first step in the data wrangling process. Ath this point the high level gathering process was finished.  
As result three DataFrames were obtained:
1. DataFrame from an existing csv-file (twitter-archive-enhanced.csv) using pandas
2. DDataFrame from tsc-file downloaded from the internet (image-predictions.tsv) using requests
3. Querying an API (tweet_json.txt) obtaining JSON object of all the tweet_ids using tweepy    

Obtained data was imported into programming environment

<a id='data_assessing'></a>
# Data Assesing

Assessing the data is the second step in data wrangling. We are going to inspect the dataset for data quality issues (i.e. content issues) and lack of tidiness (i.e. structural issues).    

We will assess data for:
1. Quality: issues with content. Low quality data is also known as dirty data.
2. Tidiness: issues with structure that prevent easy analysis. Untidy data is also known as messy data.    

Tidy data requirements:
1. Each variable forms a column.
2. Each observation forms a row. 
3. Each type of observational unit forms a table.



### Visual assessment

In [17]:
df_archive

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0     892420643555336193                    NaN                  NaN   
1     892177421306343426                    NaN                  NaN   
2     891815181378084864                    NaN                  NaN   
3     891689557279858688                    NaN                  NaN   
4     891327558926688256                    NaN                  NaN   
5     891087950875897856                    NaN                  NaN   
6     890971913173991426                    NaN                  NaN   
7     890729181411237888                    NaN                  NaN   
8     890609185150312448                    NaN                  NaN   
9     890240255349198849                    NaN                  NaN   
10    890006608113172480                    NaN                  NaN   
11    889880896479866881                    NaN                  NaN   
12    889665388333682689                    NaN                  NaN   
13    889638837579907072                    NaN                  NaN   
14    889531135344209921                    NaN                  NaN   
15    889278841981685760                    NaN                  NaN   
16    888917238123831296                    NaN                  NaN   
17    888804989199671297                    NaN                  NaN   
18    888554962724278272                    NaN                  NaN   
19    888202515573088257                    NaN                  NaN   
20    888078434458587136                    NaN                  NaN   
21    887705289381826560                    NaN                  NaN   
22    887517139158093824                    NaN                  NaN   
23    887473957103951883                    NaN                  NaN   
24    887343217045368832                    NaN                  NaN   
25    887101392804085760                    NaN                  NaN   
26    886983233522544640                    NaN                  NaN   
27    886736880519319552                    NaN                  NaN   
28    886680336477933568                    NaN                  NaN   
29    886366144734445568                    NaN                  NaN   
...                  ...                    ...                  ...   
2326  666411507551481857                    NaN                  NaN   
2327  666407126856765440                    NaN                  NaN   
2328  666396247373291520                    NaN                  NaN   
2329  666373753744588802                    NaN                  NaN   
2330  666362758909284353                    NaN                  NaN   
2331  666353288456101888                    NaN                  NaN   
2332  666345417576210432                    NaN                  NaN   
2333  666337882303524864                    NaN                  NaN   
2334  666293911632134144                    NaN                  NaN   
2335  666287406224695296                    NaN                  NaN   
2336  666273097616637952                    NaN                  NaN   
2337  666268910803644416                    NaN                  NaN   
2338  666104133288665088                    NaN                  NaN   
2339  666102155909144576                    NaN                  NaN   
2340  666099513787052032                    NaN                  NaN   
2341  666094000022159362                    NaN                  NaN   
2342  666082916733198337                    NaN                  NaN   
2343  666073100786774016                    NaN                  NaN   
2344  666071193221509120                    NaN                  NaN   
2345  666063827256086533                    NaN                  NaN   
2346  666058600524156928                    NaN                  NaN   
2347  666057090499244032                    NaN                  NaN   
2348  666055525042405380                    NaN                  NaN   
2349  666051853826850816                    NaN         

In [18]:
df_image

tweet_id                                            jpg_url  \
0     666020888022790149    https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1     666029285002620928    https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2     666033412701032449    https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3     666044226329800704    https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4     666049248165822465    https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   
5     666050758794694657    https://pbs.twimg.com/media/CT5Jof1WUAEuVxN.jpg   
6     666051853826850816    https://pbs.twimg.com/media/CT5KoJ1WoAAJash.jpg   
7     666055525042405380    https://pbs.twimg.com/media/CT5N9tpXIAAifs1.jpg   
8     666057090499244032    https://pbs.twimg.com/media/CT5PY90WoAAQGLo.jpg   
9     666058600524156928    https://pbs.twimg.com/media/CT5Qw94XAAA_2dP.jpg   
10    666063827256086533    https://pbs.twimg.com/media/CT5Vg_wXIAAXfnj.jpg   
11    666071193221509120    https://pbs.twimg.com/media/CT5cN_3WEAAlOoZ.jpg   
12    666073100786774016    https://pbs.twimg.com/media/CT5d9DZXAAALcwe.jpg   
13    666082916733198337    https://pbs.twimg.com/media/CT5m4VGWEAAtKc8.jpg   
14    666094000022159362    https://pbs.twimg.com/media/CT5w9gUW4AAsBNN.jpg   
15    666099513787052032    https://pbs.twimg.com/media/CT51-JJUEAA6hV8.jpg   
16    666102155909144576    https://pbs.twimg.com/media/CT54YGiWUAEZnoK.jpg   
17    666104133288665088    https://pbs.twimg.com/media/CT56LSZWoAAlJj2.jpg   
18    666268910803644416    https://pbs.twimg.com/media/CT8QCd1WEAADXws.jpg   
19    666273097616637952    https://pbs.twimg.com/media/CT8T1mtUwAA3aqm.jpg   
20    666287406224695296    https://pbs.twimg.com/media/CT8g3BpUEAAuFjg.jpg   
21    666293911632134144    https://pbs.twimg.com/media/CT8mx7KW4AEQu8N.jpg   
22    666337882303524864    https://pbs.twimg.com/media/CT9OwFIWEAMuRje.jpg   
23    666345417576210432    https://pbs.twimg.com/media/CT9Vn7PWoAA_ZCM.jpg   
24    666353288456101888    https://pbs.twimg.com/media/CT9cx0tUEAAhNN_.jpg   
25    666362758909284353    https://pbs.twimg.com/media/CT9lXGsUcAAyUFt.jpg   
26    666373753744588802    https://pbs.twimg.com/media/CT9vZEYWUAAlZ05.jpg   
27    666396247373291520    https://pbs.twimg.com/media/CT-D2ZHWIAA3gK1.jpg   
28    666407126856765440    https://pbs.twimg.com/media/CT-NvwmW4AAugGZ.jpg   
29    666411507551481857    https://pbs.twimg.com/media/CT-RugiWIAELEaq.jpg   
...                  ...                                                ...   
2045  886366144734445568    https://pbs.twimg.com/media/DE0BTnQUwAApKEH.jpg   
2046  886680336477933568    https://pbs.twimg.com/media/DE4fEDzWAAAyHMM.jpg   
2047  886736880519319552    https://pbs.twimg.com/media/DE5Se8FXcAAJFx4.jpg   
2048  886983233522544640    https://pbs.twimg.com/media/DE8yicJW0AAAvBJ.jpg   
2049  887101392804085760    https://pbs.twimg.com/media/DE-eAq6UwAA-jaE.jpg   
2050  887343217045368832  https://pbs.twimg.com/ext_tw_video_thumb/88734...   
2051  887473957103951883    https://pbs.twimg.com/media/DFDw2tyUQAAAFke.jpg   
2052  887517139158093824  https://pbs.twimg.com/ext_tw_video_thumb/88751...   
2053  887705289381826560    https://pbs.twimg.com/media/DFHDQBbXgAEqY7t.jpg   
2054  888078434458587136    https://pbs.twimg.com/media/DFMWn56WsAAkA7B.jpg   
2055  888202515573088257    https://pbs.twimg.com/media/DFDw2tyUQAAAFke.jpg   
2056  888554962724278272    https://pbs.twimg.com/media/DFTH_O-UQAACu20.jpg   
2057  888804989199671297    https://pbs.twimg.com/media/DFWra-3VYAA2piG.jpg   
2058  888917238123831296    https://pbs.twimg.com/media/DFYRgsOUQAARGhO.jpg   
2059  889278841981685760  https://pbs.twimg.com/ext_tw_video_thumb/88927...   
2060  889531135344209921    https://pbs.twimg.com/media/DFg_2PVW0AEHN3p.jpg   
2061  889638837579907072    https://pbs.twimg.com/media/DFihzFfXsAYGDPR.jpg   
2062  889665388333682689    https://pbs.twimg.com/media/DFi579UWsAAatzw.jpg   
2063  889880896479866881    https://pbs.twimg.com/media/DFl99B1WsA

In [19]:
df_tweet

tweet_id  favorite_count  retweet_count
0     892420643555336193           36297           7725
1     892177421306343426           31301           5718
2     891815181378084864           23570           3786
3     891689557279858688           39598           7888
4     891327558926688256           37789           8506
5     891087950875897856           19046           2851
6     890971913173991426           11093           1855
7     890729181411237888           61165          17242
8     890609185150312448           26200           3925
9     890240255349198849           29949           6718
10    890006608113172480           28827           6702
11    889880896479866881           26218           4565
12    889665388333682689           45150           9152
13    889638837579907072           25415           4108
14    889531135344209921           14214           2061
15    889278841981685760           23678           4873
16    888917238123831296           27365           4116
17    888804989199671297           24047           3881
18    888554962724278272           18561           3185
19    888078434458587136           20472           3166
20    887705289381826560           28431           4921
21    887517139158093824           43599          10771
22    887473957103951883           64638          16460
23    887343217045368832           31637           9571
24    887101392804085760           28782           5452
25    886983233522544640           32847           7018
26    886736880519319552           11271           2940
27    886680336477933568           21120           4086
28    886366144734445568           19880           2895
29    886267009285017600             111              4
...                  ...             ...            ...
2301  666411507551481857             411            301
2302  666407126856765440             100             32
2303  666396247373291520             159             76
2304  666373753744588802             174             85
2305  666362758909284353             727            520
2306  666353288456101888             198             66
2307  666345417576210432             274            128
2308  666337882303524864             182             84
2309  666293911632134144             469            321
2310  666287406224695296             138             62
2311  666273097616637952             161             73
2312  666268910803644416              96             32
2313  666104133288665088           13695           6003
2314  666102155909144576              72             11
2315  666099513787052032             143             60
2316  666094000022159362             154             67
2317  666082916733198337             103             41
2318  666073100786774016             297            147
2319  666071193221509120             136             53
2320  666063827256086533             446            198
2321  666058600524156928             105             54
2322  666057090499244032             270            126
2323  666055525042405380             410            221
2324  666051853826850816            1129            777
2325  666050758794694657             124             51
2326  666049248165822465              96             39
2327  666044226329800704             272            132
2328  666033412701032449             112             41
2329  666029285002620928             121             42
2330  666020888022790149            2418            462

[2331 rows x 3 columns]

### Programmatic assessment

### Assessing   `df_archive` DataFrame

In [20]:
df_archive.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 17 columns):
tweet_id                      2356 non-null int64
in_reply_to_status_id         78 non-null float64
in_reply_to_user_id           78 non-null float64
timestamp                     2356 non-null object
source                        2356 non-null object
text                          2356 non-null object
retweeted_status_id           181 non-null float64
retweeted_status_user_id      181 non-null float64
retweeted_status_timestamp    181 non-null object
expanded_urls                 2297 non-null object
rating_numerator              2356 non-null int64
rating_denominator            2356 non-null int64
name                          2356 non-null object
doggo                         2356 non-null object
floofer                       2356 non-null object
pupper                        2356 non-null object
puppo                         2356 non-null object
dtypes: float64(4), int64(3), ob

**Quick summary based on information above:**    
Total number of entries : 2356    
Number of retweets : 181    
Missing values in columns `in_reply_to_status_id`, `in_reply_to_user_id`, `expanded_urls`  
Data type of columns  `in_reply_to_status_id`, `in_reply_to_user_id`, `retweeted_status_id`, `retweeted_status_user_id` is float    
Data type of columns `timestamp` and `retweeted_status_timestamp` is object (string)

In [21]:
#checking if there are some duplicates (rows) in Data Frame
df_archive.duplicated().sum()

0

In [22]:
# checking for unique source values
df_archive['source'].unique()

array([ '<a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>',
       '<a href="http://twitter.com" rel="nofollow">Twitter Web Client</a>',
       '<a href="http://vine.co" rel="nofollow">Vine - Make a Scene</a>',
       '<a href="https://about.twitter.com/products/tweetdeck" rel="nofollow">TweetDeck</a>'], dtype=object)

In [23]:
# examine the column expanded_urls
df_archive['expanded_urls'].unique()

array(['https://twitter.com/dog_rates/status/892420643555336193/photo/1',
       'https://twitter.com/dog_rates/status/892177421306343426/photo/1',
       'https://twitter.com/dog_rates/status/891815181378084864/photo/1',
       ...,
       'https://twitter.com/dog_rates/status/666033412701032449/photo/1',
       'https://twitter.com/dog_rates/status/666029285002620928/photo/1',
       'https://twitter.com/dog_rates/status/666020888022790149/photo/1'], dtype=object)

In [24]:
df_archive[['rating_numerator', 'rating_denominator']].describe()

rating_numerator  rating_denominator
count       2356.000000         2356.000000
mean          13.126486           10.455433
std           45.876648            6.745237
min            0.000000            0.000000
25%           10.000000           10.000000
50%           11.000000           10.000000
75%           12.000000           10.000000
max         1776.000000          170.000000

Look like there are outliers in the columns `rating_numerator` and `rating_denominator`	

In [25]:
# examine the column rating_numerator
df_archive['rating_numerator'].value_counts().sort_index(ascending = False)

1776      1
960       1
666       1
420       2
204       1
182       1
165       1
144       1
143       1
121       1
99        1
88        1
84        1
80        1
75        2
60        1
50        1
45        1
44        1
27        1
26        1
24        1
20        1
17        1
15        2
14       54
13      351
12      558
11      464
10      461
9       158
8       102
7        55
6        32
5        37
4        17
3        19
2         9
1         9
0         2
Name: rating_numerator, dtype: int64

In [26]:
# how many entries have numerator rating greater than 15
df_archive[df_archive['rating_numerator'] > 15].shape[0]

26

Only 26 entries of 2356 with numerator rating over 15

In [27]:
# examine the column rating_denominator
df_archive['rating_denominator'].value_counts().sort_index(ascending = False)

170       1
150       1
130       1
120       1
110       1
90        1
80        2
70        1
50        3
40        1
20        2
16        1
15        1
11        3
10     2333
7         1
2         1
0         1
Name: rating_denominator, dtype: int64

In [28]:
# how many entries have denominator rating greater than 10
df_archive[df_archive['rating_denominator'] > 10].shape[0]

20

Only 20 entries of 2356 with denominator rating over 10

In [29]:
# examine the column name
df_archive['name'].value_counts()

None         745
a             55
Charlie       12
Cooper        11
Oliver        11
Lucy          11
Tucker        10
Penny         10
Lola          10
Bo             9
Winston        9
the            8
Sadie          8
an             7
Toby           7
Daisy          7
Bailey         7
Buddy          7
Leo            6
Milo           6
Scout          6
Dave           6
Jax            6
Jack           6
Bella          6
Rusty          6
Oscar          6
Stanley        6
Koda           6
Gus            5
            ... 
Jeremy         1
Filup          1
Bradlay        1
Willow         1
Alejandro      1
Halo           1
Darby          1
Rodman         1
Billl          1
Laela          1
Barclay        1
Pip            1
Batdog         1
Bode           1
Edd            1
Ralphie        1
Jockson        1
Snickers       1
Linus          1
Hermione       1
Shikha         1
Chloe          1
Bobby          1
space          1
Zeek           1
BeBe           1
Walker         1
Donny         

In [30]:
print(set([name for name in df_archive['name'] if name[0].isupper()]))

{'Molly', 'Cecil', 'Rocky', 'Thor', 'Sid', 'Nigel', 'Farfle', 'Kilo', 'Apollo', 'Timofy', 'Marvin', 'Petrick', 'Taco', 'Tyr', 'Abby', 'Emanuel', 'Kuyu', 'Ash', 'Dante', 'Olive', 'Kenneth', 'Dakota', 'Buddah', 'Astrid', 'Fletcher', 'Beemo', 'Obie', 'Atlas', 'Luther', 'Mo', 'Hermione', 'Hammond', 'Yogi', 'Herschel', 'Bobb', 'Snicku', 'Grizz', 'Striker', 'Monkey', 'Glenn', 'Scooter', 'Reggie', 'Ito', 'Reagan', 'Arlo', 'Benny', 'Pavlov', 'Jeph', 'Brandonald', 'Jordy', 'Eevee', 'Angel', 'Shnuggles', 'Rizzy', 'Dido', 'Karma', 'Chesterson', 'Stubert', 'Moreton', 'Otis', 'Joey', 'Kody', 'Walker', 'Romeo', 'Bert', 'Skittles', 'Sprout', 'Oreo', 'Bradlay', 'Cora', 'Milky', 'Jesse', 'Buckley', 'Diogi', 'Chubbs', 'Eazy', 'Arnold', 'Bronte', 'Mojo', 'Kenzie', 'Opal', 'Patch', 'Lugan', 'Noosh', 'Rocco', 'Logan', 'Walter', 'Bruiser', 'Beya', 'Harnold', 'Steve', 'Rufus', 'Rontu', 'Blue', 'Reginald', 'Mark', 'Quinn', 'Andy', 'Snoopy', 'Koko', 'Colin', 'Taz', 'Sora', 'Sansa', 'Oliviér', 'Frank', 'Chet', 

In [31]:
df_archive['name'][df_archive['name'].str.islower()].unique()

array(['such', 'a', 'quite', 'not', 'one', 'incredibly', 'mad', 'an',
       'very', 'just', 'my', 'his', 'actually', 'getting', 'this',
       'unacceptable', 'all', 'old', 'infuriating', 'the', 'by',
       'officially', 'life', 'light', 'space'], dtype=object)

### Quality issues of `df_archive` DataFrame:
- `user_id` is numeric 
- `retweeted_status_id` contains 181 retweets
- `in_reply_to_status_id`, `in_reply_to_user_id`, `retweeted_status_id`, `retweeted_status_user_id`, `retweeted_status_timestamp` contain further information abour retweets
- Missing values in `expanded_urls` 
- Column `source` is not properly formatted (HTML tags)
- Column `timestamp` has type object (string)   
- Outliers in `rating_numerator` and `rating_denominator` columns
- Column `name` has not only dogs' names but also words like `None`, `all`, `not`, `one` etc.

### Tydiness issues:
- `doggo`, `floofer`, `pupper`, `puppo` describe the same measurement unit `dog stage`

### Assessing the  `df_image` DataFrame

In [33]:
df_image.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2075 entries, 0 to 2074
Data columns (total 12 columns):
tweet_id    2075 non-null int64
jpg_url     2075 non-null object
img_num     2075 non-null int64
p1          2075 non-null object
p1_conf     2075 non-null float64
p1_dog      2075 non-null bool
p2          2075 non-null object
p2_conf     2075 non-null float64
p2_dog      2075 non-null bool
p3          2075 non-null object
p3_conf     2075 non-null float64
p3_dog      2075 non-null bool
dtypes: bool(3), float64(3), int64(2), object(4)
memory usage: 152.1+ KB


**Quick summary based on information above:**    
Total number of entries : 2074    
No missing values    
Columns' names don't clear describe the content

In [34]:
# examine the jpg_url column
df_image['jpg_url'].unique()

array(['https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg',
       'https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg',
       'https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg', ...,
       'https://pbs.twimg.com/media/DGBdLU1WsAANxJ9.jpg',
       'https://pbs.twimg.com/media/DGGmoV4XsAAUL6n.jpg',
       'https://pbs.twimg.com/media/DGKD1-bXoAAIAUK.jpg'], dtype=object)

In [35]:
# examine the img_num column
df_image['img_num'].unique()

array([1, 4, 2, 3])

In [ ]:
# examine p1
df_image['p1']

In [43]:
# are there some entries which were qualified not as dogs
for col in ['p1_dog', 'p2_dog', 'p3_dog']:
    print(f' {col} contains {df_image[col][df_image[col] == False].count()} non dog entries')

 p1_dog contains 543 non dog entries
 p2_dog contains 522 non dog entries
 p3_dog contains 576 non dog entries


### Quality issues of `df_image` DataFrame:
- `tweet_id` is numeric
- Non-dogs entries (rows where p1_dog, p2_dog, and p3_dog are False)

### Tydiness issues of `df_image` DataFrame:
- Columns' names don't clear describe the content


### Assesing `df_tweet` DataFrame

In [44]:
df_tweet.head()

tweet_id  favorite_count  retweet_count
0  892420643555336193           36297           7725
1  892177421306343426           31301           5718
2  891815181378084864           23570           3786
3  891689557279858688           39598           7888
4  891327558926688256           37789           8506

In [45]:
df_tweet.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2331 entries, 0 to 2330
Data columns (total 3 columns):
tweet_id          2331 non-null object
favorite_count    2331 non-null int64
retweet_count     2331 non-null int64
dtypes: int64(2), object(1)
memory usage: 54.7+ KB


In [46]:
df_tweet.describe()

favorite_count  retweet_count
count     2331.000000    2331.000000
mean      7580.459459    2711.719434
std      11764.976355    4586.146216
min          0.000000       1.000000
25%       1322.500000     548.500000
50%       3297.000000    1272.000000
75%       9286.000000    3151.000000
max     156665.000000   78067.000000

**Quick summary based on information above:**    
Total number of entries : 2331   
No missing values
`tweet_id` has already type object

<a id='data_cleaning'></a>
# Data Cleaning 

## Cleaning `df_image` DataFrame

**DEFINE :** changing the type of `tweet_id` from integer to string

**CODE :**

In [300]:
df_archive['tweet_id'] = df_archive['tweet_id'].astype(str)

**TEST :**

In [301]:
df_archive['tweet_id'].describe()

count                   2356
unique                  2356
top       666407126856765440
freq                       1
Name: tweet_id, dtype: object

**DEFINE :** Remove all retweets rows with `retweeted_status_*` since we are interested in original tweets only. Check to make sure the values are decreased by 181, the number of retweets (from 2356 to 2175) then drop those columns. Drop the `in_reply_to_*` columns as these are unneccessary.

**CODE :**

In [302]:
df_archive.drop(df_archive[df_archive['retweeted_status_id'].notnull()].index, inplace = True)

**TEST :**

In [303]:
df_archive.shape[0]

2175

**DEFINE :** Drop columns with retweet information. Sinse we are droping five columns the number of columns schould decrease from 17 to 12

**CODE :**

In [304]:
df_archive.drop(['retweeted_status_id',
                 'retweeted_status_user_id',
                 'retweeted_status_timestamp',
                 'in_reply_to_status_id',
                 'in_reply_to_user_id'], axis=1, inplace=True)

**TEST :**

In [305]:
df_archive.shape[1]

12

**DEFINE :** dropping rows with missing values in `expanded_urls`. Since there are 58 rows the number of rows schould decrease from 2175 to 2117

**CODE :**

In [306]:
df_archive.drop(df_archive[df_archive['expanded_urls'].isna()].index, inplace = True)

**TEST :**

In [307]:
df_archive.shape[0]

2117

**DEFINE :** Get rid of HTML code in `source`

**CODE :**

In [308]:
df_archive['source'].unique()

array([ '<a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>',
       '<a href="http://twitter.com" rel="nofollow">Twitter Web Client</a>',
       '<a href="http://vine.co" rel="nofollow">Vine - Make a Scene</a>',
       '<a href="https://about.twitter.com/products/tweetdeck" rel="nofollow">TweetDeck</a>'], dtype=object)

In [309]:
html_dict = {'<a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>': 'Twitter for iPhone',
             '<a href="http://twitter.com" rel="nofollow">Twitter Web Client</a>': 'Twitter Web Client',
             '<a href="http://vine.co" rel="nofollow">Vine - Make a Scene</a>': 'Vine - Make a Scene',
             '<a href="https://about.twitter.com/products/tweetdeck" rel="nofollow">TweetDeck</a>': 'TweetDeck'
}

In [310]:
new_source_col = []
for entrie in df_archive['source']:
    if entrie in html_dict.keys():
        new_source_col.append(html_dict[entrie])
    else:
        new_source_col.append(np.nan)

df_archive['source'] = new_source_col


**TEST :**

In [311]:
df_archive['source'].unique()

array(['Twitter for iPhone', 'Twitter Web Client', 'Vine - Make a Scene',
       'TweetDeck'], dtype=object)

In [312]:
df_archive['source'].isnull().sum()

0

**DEFINE :** converting `timestamp` to datetime object

**CODE :**

In [313]:
df_archive['timestamp'] = pd.to_datetime(df_archive['timestamp'])

**TEST :**

In [314]:
df_archive.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2117 entries, 0 to 2355
Data columns (total 12 columns):
tweet_id              2117 non-null object
timestamp             2117 non-null datetime64[ns]
source                2117 non-null object
text                  2117 non-null object
expanded_urls         2117 non-null object
rating_numerator      2117 non-null int64
rating_denominator    2117 non-null int64
name                  2117 non-null object
doggo                 2117 non-null object
floofer               2117 non-null object
pupper                2117 non-null object
puppo                 2117 non-null object
dtypes: datetime64[ns](1), int64(2), object(9)
memory usage: 215.0+ KB


**DEFINE :** dealing with outliers in `rating_numerator` und `rating_denominator`

**CODE :**

In [315]:
filt_num = df_archive['rating_numerator'] > 15

In [316]:
df_archive.loc[filt_num, 'rating_numerator'] = 15

In [317]:
df_archive['rating_numerator'].value_counts().sort_index(ascending=False)

15     19
14     40
13    295
12    489
11    417
10    436
9     154
8      98
7      52
6      32
5      34
4      16
3      19
2       9
1       5
0       2
Name: rating_numerator, dtype: int64

In [318]:
filt_den = df_archive['rating_denominator'] > 10

In [319]:
df_archive.loc[filt_den, 'rating_denominator'] = 10

**TEST :**

In [320]:
df_archive['rating_numerator'].value_counts().sort_index(ascending = False)

15     19
14     40
13    295
12    489
11    417
10    436
9     154
8      98
7      52
6      32
5      34
4      16
3      19
2       9
1       5
0       2
Name: rating_numerator, dtype: int64

In [321]:
df_archive['rating_denominator'].value_counts()

10    2115
7        1
2        1
Name: rating_denominator, dtype: int64

**DEFINE :** get rid of all non-dog name in `name`

**CODE :**

In [322]:
not_dog_names = df_archive['name'][df_archive['name'].str.islower()].unique()

In [323]:
new_dog_names = []
for name in df_archive['name']:
    if name in not_dog_names or name == 'None':
        new_dog_names.append(np.nan)
    else:
        new_dog_names.append(name)
        

In [324]:
df_archive['name'] = new_dog_names

In [325]:
df_archive['name'].isnull().sum()

726

In [326]:
df_archive.drop(df_archive[df_archive['name'].isna()].index, inplace = True)

**TEST :**

In [327]:
df_archive.shape

(1391, 12)

**DEFINE :** Combine dog stage columns `doggo`, `floofer`, `pupper`, `puppo` into one `dog_stage` column. Delete the separate dog stage categories. Convert entries in this column title case

**CODE :**

In [328]:
df_archive.head()

tweet_id           timestamp              source  \
0  892420643555336193 2017-08-01 16:23:56  Twitter for iPhone   
1  892177421306343426 2017-08-01 00:17:27  Twitter for iPhone   
2  891815181378084864 2017-07-31 00:18:03  Twitter for iPhone   
3  891689557279858688 2017-07-30 15:58:51  Twitter for iPhone   
4  891327558926688256 2017-07-29 16:00:24  Twitter for iPhone   

                                                text  \
0  This is Phineas. He's a mystical boy. Only eve...   
1  This is Tilly. She's just checking pup on you....   
2  This is Archie. He is a rare Norwegian Pouncin...   
3  This is Darla. She commenced a snooze mid meal...   
4  This is Franklin. He would like you to stop ca...   

                                       expanded_urls  rating_numerator  \
0  https://twitter.com/dog_rates/status/892420643...                13   
1  https://twitter.com/dog_rates/status/892177421...                13   
2  https://twitter.com/dog_rates/status/891815181...                12   
3  https://twitter.com/dog_rates/status/891689557...                13   
4  https://twitter.com/dog_rates/status/891327558...                12   

   rating_denominator      name doggo floofer pupper puppo  
0                  10   Phineas  None    None   None  None  
1                  10     Tilly  None    None   None  None  
2                  10    Archie  None    None   None  None  
3                  10     Darla  None    None   None  None  
4                  10  Franklin  None    None   None  None

In [329]:
df_archive['doggo'].unique()

array(['None', 'doggo'], dtype=object)

In [330]:
df_archive['floofer'].unique()

array(['None', 'floofer'], dtype=object)

In [331]:
df_archive['pupper'].unique()

array(['None', 'pupper'], dtype=object)

In [332]:
df_archive['puppo'].unique()

array(['None', 'puppo'], dtype=object)

In [333]:
for stage in ['doggo','floofer', 'pupper','puppo']:
    df_archive[stage].replace('None', '', inplace = True)

In [334]:
df_archive['dog_stage'] = df_archive['doggo'] + df_archive['floofer'] + df_archive['pupper'] + df_archive['puppo']

In [335]:
df_archive['dog_stage'].unique()

array(['', 'doggo', 'puppo', 'pupper', 'floofer', 'doggopupper'], dtype=object)

In [336]:
df_archive['dog_stage'][df_archive['dog_stage'] == ''].count()

1207

In [337]:
df_archive[stage].replace('', 'Unknown', inplace = True)

In [338]:
df_archive['dog_stage'].unique()

array(['', 'doggo', 'puppo', 'pupper', 'floofer', 'doggopupper'], dtype=object)

In [340]:
df_archive['dog_stage'][df_archive['dog_stage'] == 'doggopupper']

460    doggopupper
575    doggopupper
705    doggopupper
889    doggopupper
Name: dog_stage, dtype: object

In [342]:
df_archive.drop(df_archive['dog_stage'][df_archive['dog_stage'] == 'doggopupper'].index, inplace = True)

In [343]:
df_archive['dog_stage'].unique()

array(['', 'doggo', 'puppo', 'pupper', 'floofer'], dtype=object)

In [346]:
df_archive.loc[df_archive['dog_stage'] == '', 'dog_stage'] = 'Unknown'

In [344]:
df_archive.drop(['doggo', 'floofer', 'pupper', 'puppo'], axis = 1, inplace = True)

In [350]:
df_archive['dog_stage'] = df_archive['dog_stage'].str.capitalize()

**TEST :**

In [351]:
df_archive['dog_stage'].unique()

array(['Unknown', 'Doggo', 'Puppo', 'Pupper', 'Floofer'], dtype=object)

Make a clean copy of `df_archive` DataFrame

In [353]:
df_archive_clean = df_archive.copy()

## Cleaning `df_image` DataFrame

In [375]:
df_image.head()

tweet_id                                          jpg_url  \
0  666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1  666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2  666033412701032449  https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3  666044226329800704  https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4  666049248165822465  https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   

   img_num                      p1   p1_conf  p1_dog                  p2  \
0        1  Welsh_springer_spaniel  0.465074    True              collie   
1        1                 redbone  0.506826    True  miniature_pinscher   
2        1         German_shepherd  0.596461    True            malinois   
3        1     Rhodesian_ridgeback  0.408143    True             redbone   
4        1      miniature_pinscher  0.560311    True          Rottweiler   

    p2_conf  p2_dog                   p3   p3_conf  p3_dog  
0  0.156665    True    Shetland_sheepdog  0.061428    True  
1  0.074192    True  Rhodesian_ridgeback  0.072010    True  
2  0.138584    True           bloodhound  0.116197    True  
3  0.360687    True   miniature_pinscher  0.222752    True  
4  0.243682    True             Doberman  0.154629    True

In [376]:
col_names = ['tweet_id', 'jpg_url', 'img_num', 
             'breed1', 'breed1_prob', 'breed1_dog', 
             'breed2', 'breed2_prob', 'breed2_dog',
             'breed3', 'breed3_prob', 'breed3_dog',]
df_image.columns = col_names
df_image.head()

tweet_id                                          jpg_url  \
0  666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1  666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2  666033412701032449  https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3  666044226329800704  https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4  666049248165822465  https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   

   img_num                  breed1  breed1_prob  breed1_dog  \
0        1  Welsh_springer_spaniel     0.465074        True   
1        1                 redbone     0.506826        True   
2        1         German_shepherd     0.596461        True   
3        1     Rhodesian_ridgeback     0.408143        True   
4        1      miniature_pinscher     0.560311        True   

               breed2  breed2_prob  breed2_dog               breed3  \
0              collie     0.156665        True    Shetland_sheepdog   
1  miniature_pinscher     0.074192        True  Rhodesian_ridgeback   
2            malinois     0.138584        True           bloodhound   
3             redbone     0.360687        True   miniature_pinscher   
4          Rottweiler     0.243682        True             Doberman   

   breed3_prob  breed3_dog  
0     0.061428        True  
1     0.072010        True  
2     0.116197        True  
3     0.222752        True  
4     0.154629        True

In [377]:
df_image['dog_pred'] =  df_image['breed1_dog'].astype(int) \
                        + df_image['breed2_dog'].astype(int) \
                        + df_image['breed3_dog'].astype(int)

In [378]:
number = [0,1,2,3]
dog = ['Not Dog', 'Maybe Dog', 'Maybe Dog', 'Dog']
for num, pred in zip(number, dog):
    df_image['dog_pred'] = df_image['dog_pred'].replace(num, pred)

In [379]:
df_image['dog_pred'].unique()

array(['Dog', 'Not Dog', 'Maybe Dog'], dtype=object)

In [363]:
for breed in ['p1', 'p2', 'p3']:
    junk_words = df_image[breed][df_image[breed].str.islower()]
    new_dog_names = []
    for name in df_image[breed]:
        if name not in junk_words:
            splitted_name = name.split('_')
            new_name = ' '.join(splitted_name).title()
        else:
            new_name = np.nan
        new_dog_names.append(new_name)
    df_image[breed] = new_dog_names

In [364]:
df_image

tweet_id                                            jpg_url  \
0     666020888022790149    https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1     666029285002620928    https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2     666033412701032449    https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3     666044226329800704    https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4     666049248165822465    https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   
5     666050758794694657    https://pbs.twimg.com/media/CT5Jof1WUAEuVxN.jpg   
6     666051853826850816    https://pbs.twimg.com/media/CT5KoJ1WoAAJash.jpg   
7     666055525042405380    https://pbs.twimg.com/media/CT5N9tpXIAAifs1.jpg   
8     666057090499244032    https://pbs.twimg.com/media/CT5PY90WoAAQGLo.jpg   
9     666058600524156928    https://pbs.twimg.com/media/CT5Qw94XAAA_2dP.jpg   
10    666063827256086533    https://pbs.twimg.com/media/CT5Vg_wXIAAXfnj.jpg   
11    666071193221509120    https://pbs.twimg.com/media/CT5cN_3WEAAlOoZ.jpg   
12    666073100786774016    https://pbs.twimg.com/media/CT5d9DZXAAALcwe.jpg   
13    666082916733198337    https://pbs.twimg.com/media/CT5m4VGWEAAtKc8.jpg   
14    666094000022159362    https://pbs.twimg.com/media/CT5w9gUW4AAsBNN.jpg   
15    666099513787052032    https://pbs.twimg.com/media/CT51-JJUEAA6hV8.jpg   
16    666102155909144576    https://pbs.twimg.com/media/CT54YGiWUAEZnoK.jpg   
17    666104133288665088    https://pbs.twimg.com/media/CT56LSZWoAAlJj2.jpg   
18    666268910803644416    https://pbs.twimg.com/media/CT8QCd1WEAADXws.jpg   
19    666273097616637952    https://pbs.twimg.com/media/CT8T1mtUwAA3aqm.jpg   
20    666287406224695296    https://pbs.twimg.com/media/CT8g3BpUEAAuFjg.jpg   
21    666293911632134144    https://pbs.twimg.com/media/CT8mx7KW4AEQu8N.jpg   
22    666337882303524864    https://pbs.twimg.com/media/CT9OwFIWEAMuRje.jpg   
23    666345417576210432    https://pbs.twimg.com/media/CT9Vn7PWoAA_ZCM.jpg   
24    666353288456101888    https://pbs.twimg.com/media/CT9cx0tUEAAhNN_.jpg   
25    666362758909284353    https://pbs.twimg.com/media/CT9lXGsUcAAyUFt.jpg   
26    666373753744588802    https://pbs.twimg.com/media/CT9vZEYWUAAlZ05.jpg   
27    666396247373291520    https://pbs.twimg.com/media/CT-D2ZHWIAA3gK1.jpg   
28    666407126856765440    https://pbs.twimg.com/media/CT-NvwmW4AAugGZ.jpg   
29    666411507551481857    https://pbs.twimg.com/media/CT-RugiWIAELEaq.jpg   
...                  ...                                                ...   
2045  886366144734445568    https://pbs.twimg.com/media/DE0BTnQUwAApKEH.jpg   
2046  886680336477933568    https://pbs.twimg.com/media/DE4fEDzWAAAyHMM.jpg   
2047  886736880519319552    https://pbs.twimg.com/media/DE5Se8FXcAAJFx4.jpg   
2048  886983233522544640    https://pbs.twimg.com/media/DE8yicJW0AAAvBJ.jpg   
2049  887101392804085760    https://pbs.twimg.com/media/DE-eAq6UwAA-jaE.jpg   
2050  887343217045368832  https://pbs.twimg.com/ext_tw_video_thumb/88734...   
2051  887473957103951883    https://pbs.twimg.com/media/DFDw2tyUQAAAFke.jpg   
2052  887517139158093824  https://pbs.twimg.com/ext_tw_video_thumb/88751...   
2053  887705289381826560    https://pbs.twimg.com/media/DFHDQBbXgAEqY7t.jpg   
2054  888078434458587136    https://pbs.twimg.com/media/DFMWn56WsAAkA7B.jpg   
2055  888202515573088257    https://pbs.twimg.com/media/DFDw2tyUQAAAFke.jpg   
2056  888554962724278272    https://pbs.twimg.com/media/DFTH_O-UQAACu20.jpg   
2057  888804989199671297    https://pbs.twimg.com/media/DFWra-3VYAA2piG.jpg   
2058  888917238123831296    https://pbs.twimg.com/media/DFYRgsOUQAARGhO.jpg   
2059  889278841981685760  https://pbs.twimg.com/ext_tw_video_thumb/88927...   
2060  889531135344209921    https://pbs.twimg.com/media/DFg_2PVW0AEHN3p.jpg   
2061  889638837579907072    https://pbs.twimg.com/media/DFihzFfXsAYGDPR.jpg   
2062  889665388333682689    https://pbs.twimg.com/media/DFi579UWsAAatzw.jpg   
2063  889880896479866881    https://pbs.twimg.com/media/DFl99B1WsA